storage: use map for watchableStorage.unsynced

gyuho opened the pull request:
From `TODO: use map to reduce cancel cost`:
This changes unsynced from `[]*watcher` to `map[*watcher]struct{}`,
and fixes a line that seemed to be missing `defer` statement.

Here's the benchmark results:

```
With []*watcher:

BenchmarkStorePut	   20000	     63670 ns/op	    2120 B/op	      26 allocs/op
BenchmarkStorePut-2	   30000	     34961 ns/op	    1885 B/op	      22 allocs/op
BenchmarkStorePut-4	   30000	     38461 ns/op	    1862 B/op	      22 allocs/op
BenchmarkStorePut-8	   30000	     35052 ns/op	    1896 B/op	      22 allocs/op
BenchmarkStorePut-16	   30000	     34053 ns/op	    1916 B/op	      22 allocs/op
BenchmarkKVWatcherMemoryUsage	  200000	     22775 ns/op	     614 B/op	      13 allocs/op
BenchmarkKVWatcherMemoryUsage-2	  500000	      2791 ns/op	     676 B/op	      13 allocs/op
BenchmarkKVWatcherMemoryUsage-4	  500000	      2269 ns/op	     676 B/op	      13 allocs/op
BenchmarkKVWatcherMemoryUsage-8	  500000	      2209 ns/op	     676 B/op	      13 allocs/op
BenchmarkKVWatcherMemoryUsage-16	  500000	      2108 ns/op	     676 B/op	      13 allocs/op

With map[*watcher]struct{}:

BenchmarkStorePut	   30000	     54057 ns/op	    1777 B/op	      21 allocs/op
BenchmarkStorePut-2	   30000	     38730 ns/op	    1718 B/op	      20 allocs/op
BenchmarkStorePut-4	   30000	     41090 ns/op	    1892 B/op	      22 allocs/op
BenchmarkStorePut-8	   20000	     52318 ns/op	    1972 B/op	      23 allocs/op
BenchmarkStorePut-16	   50000	     32132 ns/op	    1545 B/op	      16 allocs/op
BenchmarkKVWatcherMemoryUsage	  100000	     12495 ns/op	     599 B/op	      13 allocs/op
BenchmarkKVWatcherMemoryUsage-2	  500000	      3046 ns/op	     677 B/op	      13 allocs/op
BenchmarkKVWatcherMemoryUsage-4	  500000	      2214 ns/op	     676 B/op	      13 allocs/op
BenchmarkKVWatcherMemoryUsage-8	  500000	      2159 ns/op	     676 B/op	      13 allocs/op
BenchmarkKVWatcherMemoryUsage-16	  500000	      2121 ns/op	     676 B/op	      13 allocs/op

```

Performance gain is unnoticeable.

Thanks,

yichengq commented:
Could you send `defer` one in a separate PR first? So this makes the goal of PR clear.

Several questions for the PR:
1. the original code in BenchmarkKVWatcherMemoryUsage will let watcher go into `synced` map instead of `unsynced`. So I think your benchmark is not useful.
2. The execution speed in two benchmarks changes much in `BenchmarkKVWatcherMemoryUsage`. Do you know why it happens?

gyuho commented:
@yichengq Thanks for the review.

I agree. I will be more complete and detailed in the next
commit addressing the issues you mentioned. I will come
back to this tonight.

Thanks,

gyuho commented:
Closing this in favor of https://github.com/coreos/etcd/pull/3577 and
another pull request coming with more complete benchmark.

I would like to separate as @yichengq suggested:
- fix for `defer Unlock`
- PR for using map

gyuho commented:
Another related PR: https://github.com/coreos/etcd/pull/3588
